### PR TITLE
auth: identity token source

### PR DESCRIPTION
**Pull request by gwik**
Adds a token source to fetch a JWT token from the metadata server for a specific audience.

See:

https://cloud.google.com/run/docs/authenticating/service-to-service#use_the_metadata_server

### CODE REVIEW

File: foundation/auth/src/project.rs [hunk 1]
```diff
@@ -93,11 +94,22 @@ pub async fn create_token_source_from_project(
...
-        }
+        Project::FromMetadataServer(_) => create_token_source_from_metadata_server(&config).await,
+    }
+}
+
+/// Create a token source that use the metadata server to generate tokens.
+pub async fn create_token_source_from_metadata_server(
+    config: &Config<'_>,
+) -> Result<Box<dyn TokenSource>, error::Error> {
+    if let Some(audience) = config.audience {
```
Comment (yoshidan) [inline foundation/auth/src/project.rs:105]: The current implementation always uses ComputeIdentitySource when used via [google-cloud-default](https://github.com/golang/oauth2/blob/2323c81c8dba82e8650ed3a24a1a5667e293af38/google/default.go#L164).
If you look at the [Go library](https://github.com/golang/oauth2/blob/master/google/default.go#L160) , `ComputeTokenSource` is used in the GCP environment, so we would like to have a special option to specify `ComputeIdentitySource`. For example, by adding a flag to the Config as follows.
```rust
 if let Some(audience) = config.audience && config.use_identity_token_source {
 // use ComputeIdentitySource
} else {
  // use ComputeTokenSource
}

```

**Review by yoshidan — CHANGES_REQUESTED** [review]

File: foundation/auth/src/project.rs (hunk 1 above)
Comment (gwik) [reply foundation/auth/src/project.rs:105]: I hadn't realize it was modeled after the go library. In that case, I think it's better that the logic for "automatic" source selection stays close to the original implementation.

I propose we don't change `create_token_source_from_project` and instead we provide a top module level constructor function that one can use explicitly. In the same spirit as this:

https://github.com/golang/oauth2/blob/2323c81c8dba82e8650ed3a24a1a5667e293af38/google/jwt.go#L27

What do you think ?

Comment (gwik) [reply foundation/auth/src/project.rs:105]: Actually the idtoken library, which purpose is to call other services, [uses the `identity` endpoint](https://github.com/googleapis/google-api-go-client/blob/f06cac5563b71f297a8ee726382fadc106995924/idtoken/compute.go#L42). Maybe we should add an `idtoken` module that implement the same [selection token source logic from the credientials](https://github.com/googleapis/google-api-go-client/blob/f06cac5563b71f297a8ee726382fadc106995924/idtoken/idtoken.go#L101) ?

**Comment by antoninfin** [comment]
Here's a modified version based on the `idtoken` go package.

I'll document and test it a little bit.

**Review by yoshidan — APPROVED** [review]

**Pull request merged**